src/lm/mph_language_model.cpp

```diff
@@ -0,0 +1,411 @@
+                    " > Reading " + std::to_string(order_ + 2) + "-grams: ",
+                    options.num_keys);
+            }
+        }
+        // last
+        else
+        {
+            assert(last_builder_);
+            last_builder_->write();
+            last_builder_ = nullptr;
+        }
+    }
+
+  private:
+    std::string prefix_;
+    uint64_t order_ = 0;
+    uint64_t observed_ = 0;
+    std::vector<uint64_t> counts_;
+    std::unique_ptr<printing::progress> progress_;
+    std::unique_ptr<unigram_builder_type> unigram_builder_;
+    std::unique_ptr<ngram_map<util::string_view>> unigrams_;
+    std::unique_ptr<middle_builder_type> middle_builder_;
+    std::unique_ptr<last_builder_type> last_builder_;
+};
+
+uint64_t build_from_arpa(const std::string& arpa_file,
+                         const std::string& prefix)
+{
+    filesystem::remove_all(prefix);
+    filesystem::make_directory(prefix);
+    ngram_handler handler{prefix};
+    read_from_arpa(arpa_file, handler);
+
+    // finish off the n-grams
+    handler.finish_order();
+
+    return handler.order();
+}
+}
+
+struct mph_language_model::impl
+{
+    using unigram_map_type = ngram_map<util::string_view>;
+    using middle_map_type = ngram_map<std::vector<uint64_t>>;
+    using last_map_type = ngram_map<std::vector<uint64_t>, float>;
+
+    impl(const std::string& prefix, uint64_t o)
+        : order{o},
+          unigrams{prefix + "/0"},
+          last{prefix + "/" + std::to_string(order)}
+    {
+        for (uint64_t i = 1; i < order; ++i)
+        {
+            auto mid = make_unique<middle_map_type>(prefix + "/"
+                                                    + std::to_string(i));
+            middle_vec.push_back(std::move(mid));
+        }
+
+        unk = *unigrams.index_and_value("<unk>");
+    }
+
+    const middle_map_type& middle(uint64_t len) const
+    {
+        return *middle_vec[len - 2];
+    }
+
+    uint64_t order;
+    hashing::indexed_value<prob_backoff<>> unk;
+    unigram_map_type unigrams;
+    std::vector<std::unique_ptr<middle_map_type>> middle_vec;
+    last_map_type last;
+};
+
+mph_language_model::mph_language_model(const cpptoml::table& config)
+{
+    auto table = config.get_table("mph-language-model");
+    auto arpa_file = table->get_as<std::string>("arpa-file");
+    auto prefix = table->get_as<std::string>("binary-file-prefix");
+
+    util::optional<uint64_t> order;
+    if (!filesystem::file_exists(*prefix + "/0/values.bin"))
+    {
+        LOG(info) << "Building language model from .arpa file: " << *arpa_file
+                  << ENDLG;
+
+        auto time = common::time(
+            [&]() { order = build_from_arpa(*arpa_file, *prefix); });
+        LOG(info) << "Done. (" << time.count() << "ms)" << ENDLG;
+    }
+    LOG(info) << "Loading language model from binary files in: " << *prefix
+              << ENDLG;
+
+    if (!order)
+    {
+        for (uint64_t o = 0; filesystem::file_exists(
```

Comment:
Minor: maybe use different variable name than `o`? Appears in impl constructor as well.